core/src/mast/serialization/basic_block_data_decoder.rs

```diff
@@ -0,0 +1,244 @@
+use crate::{
+    AdviceInjector, AssemblyOp, DebugOptions, Decorator, DecoratorList, Operation, OperationData,
+    SignatureKind,
+};
+
+use super::{decorator::EncodedDecoratorVariant, StringIndex, StringRef};
+use alloc::{string::String, vec::Vec};
+use miden_crypto::{Felt, ZERO};
+use winter_utils::{ByteReader, DeserializationError, SliceReader};
+
+pub struct BasicBlockDataDecoder<'a> {
+    data: &'a [u8],
+    data_reader: SliceReader<'a>,
+    strings: &'a [StringRef],
+}
+
+/// Constructors
+impl<'a> BasicBlockDataDecoder<'a> {
+    pub fn new(data: &'a [u8], strings: &'a [StringRef]) -> Self {
+        let data_reader = SliceReader::new(data);
+
+        Self {
+            data,
+            data_reader,
+            strings,
+        }
+    }
+}
+
+/// Mutators
+impl<'a> BasicBlockDataDecoder<'a> {
+    pub fn decode_operations_and_decorators(
```

Comment:
>Question: why not implement Serializable and Deserializable for Operation instead of having separate functions here? It seems to me that encapsulating this logic there would be cleaner.

Oh yes much cleaner that way indeed. Fixed.

> Though, maybe we can still implement these traits for some of the internal decorator types (e.g., AdviceInjector, Debug, Event etc. - with AssemblyOp being the only exception).

This would be incompatible with my current strategy that "linearizes" the `Decorator` enums into one big enum `EncodedDecoratorVariant`. It would be a bigger change, and would still be messy, so I don't think it's worth it?